docs/architecture/adr-042-blockchain-riri-org.md

```diff
@@ -0,0 +1,292 @@
+
+The reactor will include a demultiplexing routine which will send each message to each sub routine for independant processing. Each sub routine will then select the messages it's interested in and call the handle specific function specified in [ADR-40](https://github.com/tendermint/tendermint/blob/develop/docs/architecture/adr-040-blockchain-reactor-refactor.md).
+
+```go
+func demuxRoutine(msgs, scheduleMsgs, processorMsgs, ioMsgs) {
+    timer := time.NewTicker(interval)
+    for {
+        select {
+            case <-timer.C:
+                now := evTimeCheck{time.Now()}
+                schedulerMsgs <- now
+                processorMsgs <- now
+                ioMsgs <- now
+            case msg:= <- msgs:
+                schedulerMsgs <- msg
+                processorMsgs <- msg
+                ioMesgs <- msg
+                if msg == stop {
+                    break;
+                }
+        }
+    }
+}
+
+func processRoutine(input chan Message, output chan Message) {
+    processor := NewProcessor(..)
+    for {
+        msg := <- input
+        switch msg := msg.(type) {
+            case bcBlockRequestMessage:
+                output <- processor.handleBlockRequest(msg, time.Now())
+            ...
+            case stop:
+                processor.stop()
+                break;
+    }
+}
+
+func scheduleRoutine(input chan Message, output chan Message) {
+    schelduer = NewScheduler(...)
+    for {
+        msg := <-msgs
+        switch msg := input.(type) {
+            case bcBlockResponseMessage:
+                output <- scheduler.handleBlockResponse(msg, time.Now())
+            ...
+            case stop:
+                schedule.stop()
+                break;
+        }
+    }
+}
+```
+
+## Lifecycle management
+
+A set of routines for individual processes allow processes to run in parallel with clear lifecycle management. `Start`, `Stop`, and `AddPeer` hooks currently present in the reactor will delegate to the sub-routines allowing them to manage internal state independant without further coupling to the reactor.
+
+```go
+func (r *BlockChainReactor) Start() {
+    r.msgs := make(chan Message, maxInFlight)
+    schedulerMsgs := make(chan Message)
+    processorMsgs := make(chan Message)
+    ioMsgs := make(chan Message)
+
+    go processorRoutine(processorMsgs, r.msgs)
+    go scheduleRoutine(schedulerMsgs, r.msgs)
+    go ioRoutine(ioMsgs, r.msgs)
+    ...
+}
+
+...
+func (r *BlockchainReactor) Stop() {
+    ...
+    r.msgs <- stop
+    ...
+}
+...
+
+func (r *BlockchainReactor) AddPeer(peer p2p.Peer) {
+    ...
+    r.msgs <- bcAddPeerEv{peer.ID}
+    ...
+}
+
+```
+
+## IO handling
+An io handling routine within the reactor will isolate peer communication.
+
+```go
+func (r *BlockchainReacor) ioRoutine(chan ioMsgs, ...) {
+    ...
+    for {
+        msg := <-ioMsgs
+        switch msg := msg.(type) {
+            case scBlockRequestMessage:
+                r.sendBlockToPeer(...)
```

Comment:
what if IO is slow? won't this slow down `demuxRoutine` in effect?

Reply:
IIRC this goes down results in a `peer.TrySend` which are queued and sent later until the queue fills up. Similar to the comment above, we can use the return value from trySend to send `stopScheduling` messages to the `scheduler` until the queue flushes.